### PR TITLE
Dev contiguous view ops

### PR DESCRIPTION
**Pull request by Flowingsun007**
本pr主要基于view/stride机制，完成以下支持view相关op的开发工作。

### 开发进度表：
| torch view op | oneflow op | view op开发进度&&pr | 认领人  |
| ------------- | ---------- | -------------- | ----------- |
| basic slice   | slice      | 已完成#6854  | 赵露阳 |
| narrow        | narrow     | 已完成#6854 | 赵露阳 |
| permute       | permute   | 已完成#6854 | 赵露阳 |
| transpose | transpose | 已完成#6854  | 赵露阳 |
| view | view  | 已完成#6854  | 赵露阳 |
| reshape | reshape | 已完成#6854 | 赵露阳 |
| expand  | expand | 已完成#7162 | 陈博文    |
| expand_as  | expand_as | 已完成#7162 |  陈博文  |
| unsqueeze | unsqueeze | 已完成#7162 |  陈博文  |
| squeeze | squeeze | 已完成#7162 | 陈博文  |
| split   | split | 已完成 (依赖narrow) | 陈博文  |
| chunk   | chunk | 已完成(在functor中通过narrow实现) | 赵露阳 |
| unfold | unfold | 已完成#6854 |   赵露阳     |
| unflatten | unflatten | ~~NOTE:该op在oneflow中尚未实现~~ |  赵露阳      |
| diagonal   | diagonal      | 已完成#6854  | 赵露阳  |
| detach        | detach     | ~~目前tensor.detach的实现已经是共享内存的，不需要经过view~~ |  赵露阳      |
| movedim        | movedim     | 已完成（在functor通过transpose实现) |  赵露阳      |
| as_strided      | as_strided   | 已完成#6854  | 马诚宇    |
| select        | select     | 当前的实现已支持view | 马诚宇    |
| unbind        | unbind     | #7730 | 马诚宇    |
| swapaxes        | swapaxes     | 当前的实现已支持view | 马诚宇    |
| swapdims        | swapdims    | #7659 | 马诚宇    |
| t        | t     | 当前的实现已支持view | 马诚宇    |
| T        | T   | 当前的实现已支持view | 马诚宇    |
| hsplit        | hsplit     | 当前的实现已支持view | 马诚宇    |
| vsplit        | vsplit     | 当前的实现已支持view | 马诚宇    |
| tensor_split        | tensor_split     | 当前的实现已支持view | 马诚宇    |
| view_as        | view_as     | #7761 | 马诚宇    |
| split_with_sizes        | split_with_sizes     | #7761 | 马诚宇    |

### oneflow尚未支持的op

| torch view op | oneflow op |
| ------------- | ---------- |
| real        | —     |
| imag        | —     |
| view_as_real        | —     |

### view相关pr:
- https://github.com/Oneflow-Inc/oneflow/pull/6815
- https://github.com/Oneflow-Inc/oneflow/pull/6852
- https://github.com/Oneflow-Inc/oneflow/pull/7158
- https://github.com/Oneflow-Inc/oneflow/pull/7162
- https://github.com/Oneflow-Inc/oneflow/pull/7147
- https://github.com/Oneflow-Inc/oneflow/pull/7503
- https://github.com/Oneflow-Inc/oneflow/pull/7719
- https://github.com/Oneflow-Inc/oneflow/pull/7670
- https://github.com/Oneflow-Inc/oneflow/pull/7776

### view相关issue：
- https://github.com/Oneflow-Inc/OneTeam/issues/472
- https://github.com/Oneflow-Inc/OneTeam/issues/776
- https://github.com/Oneflow-Inc/OneTeam/issues/1058

### 工作计划&&推进
- https://github.com/Oneflow-Inc/OneTeam/issues/776#issuecomment-999281851

### CODE REVIEW

File: oneflow/core/framework/tensor_methods.cpp [hunk 1]
```diff
@@ -98,19 +104,19 @@ Maybe<Tensor> Reshape(const std::shared_ptr<Tensor>& input, const Shape& shape)
       count *= shape.At(i);
     }
   }
-
+  int64_t storage_offset = JUST(JUST(input->AsMirroredTensor())->storage_offset());
```
Comment (hjchen2) [inline oneflow/core/framework/tensor_methods.cpp:107]: 实际上这里的Reshape对应了Pytorch的view，所以这里需要加一个判断，要求输入input必须是contiguous的

File: oneflow/core/framework/tensor_methods.cpp [hunk 2]
```diff
@@ -404,6 +404,37 @@ Maybe<Tensor> Narrow(const std::shared_ptr<Tensor>& input, const int64_t& dim, c
...
+                        const std::vector<int32_t>& size, const std::vector<int32_t>& stride,
+                        const int32_t& storage_offset) {
+  DimVector dim_vec;
+  dim_vec.insert(dim_vec.end(), size.begin(), size.end());
+  Shape target_shape(dim_vec);
+  StrideVector stride_vec(stride.size());
+  for (int i = 0; i < stride.size(); ++i) {
+    stride_vec[i] = stride[i];
+  }
+  auto output = JUST(view::BasicView(input, target_shape, Stride(stride_vec),  storage_offset));
```
Comment (Flowingsun007) [inline oneflow/core/framework/tensor_methods.cpp]: 这里传进来的storage_offset是output的，取view时是不是还需要加上原input tensor的storage_offset？

File: oneflow/core/framework/tensor_methods.cpp [hunk 3]
```diff
@@ -404,6 +404,37 @@ Maybe<Tensor> Narrow(const std::shared_ptr<Tensor>& input, const int64_t& dim, c
...
+  if (input->requires_grad()) {
+    auto backward_fn =
+        std::make_shared<std::function<Maybe<void>(const TensorTuple&, TensorTuple*, bool)>>(
+            [=](const TensorTuple& out_grads, TensorTuple* in_grads,
+                bool create_graph) -> Maybe<void> {
+              autograd::AutoGradMode mode(create_graph);
+              CHECK_EQ_OR_RETURN(out_grads.size(), 1);
+              auto like = JUST(functional::Empty(Shape(input->shape()->dim_vec()), input->dtype(),
```
Comment (Flowingsun007) [inline oneflow/core/framework/tensor_methods.cpp:422]: 这个like后面好像没用到

**Comment by Flowingsun007** [comment]
此pr已拆分为：#7670 #7503 #7813 ，且均已合入master
